### PR TITLE
Migrates publishFeature to permission class.

### PR DESCRIPTION
**Pull request by mknowlton89**
### Features and Changes

This PR migrates the `publishFeatures` to the permissions class and adds a helper `canPublishFeatures` to make it easy to pass in the experiment, and the env filter, to see if the user has permission to publish a feature.

### CODE REVIEW

**Review by jdorn — APPROVED** [review]

File: packages/back-end/src/api/features/updateFeature.ts [hunk 1]
```diff
@@ -92,17 +94,22 @@ export const updateFeature = createApiRequestHandler(updateFeatureValidator)(
...
-        updates.project,
-        getEnabledEnvironments(
-          {
-            ...feature,
-            ...updates,
-          },
-          orgEnvs
+      if (
+        !req.context.permissions.canPublishFeature(
+          feature,
```
Comment (mknowlton89) [inline packages/back-end/src/api/features/updateFeature.ts:99]: should this be `updates`?

File: packages/back-end/src/controllers/features.ts [hunk 2]
```diff
@@ -756,17 +758,27 @@ export async function postFeaturePublish(
...
   }
   // Otherwise, only the environments with rule changes are affected
   else {
     const changedEnvs = Object.keys(mergeResult.result.rules || {});
     if (changedEnvs.length > 0) {
-      req.checkPermissions("publishFeatures", feature.project, changedEnvs);
+      if (
+        !context.permissions.canPublishFeature(
+          feature,
+          Array.from(getEnabledEnvironments(feature, changedEnvs))
```
Comment (jdorn) [inline packages/back-end/src/controllers/features.ts:777]: Is this change intentional?  Went from just using `changedEnvs` to using `getEnabledEnvironments(feature, changedEnvs)`

File: packages/back-end/src/controllers/features.ts [hunk 3]
```diff
@@ -1892,14 +1913,16 @@ export async function postFeatureArchive(
   const environments = filterEnvironmentsByFeature(allEnvironments, feature);
   const environmentsIds = environments.map((e) => e.id);
 
-  if (!context.permissions.canUpdateFeature(feature, {})) {
+  if (
+    !context.permissions.canUpdateFeature(feature, {}) ||
+    !!context.permissions.canPublishFeature(
```
Comment (jdorn) [inline packages/back-end/src/controllers/features.ts:1918]: Typo. Should be `!` instead of `!!`

**Review by jdorn — APPROVED** [review]
Left 2 comments to fix

File: packages/back-end/src/controllers/features.ts (hunk 2 above)
Comment (mknowlton89) [reply packages/back-end/src/controllers/features.ts:786]: No - good catch.
